### PR TITLE
PS-7308 : Data Masking Plugin: gen_dictionary_load() doesn't check se…

### PR DESCRIPTION
**Pull request by satya-bodapati**
…cure-file-priv

Problem:
--------
gen_dictionary_load() loads a file from first argument as dictionary.
Currently this file can be located anywhere, posing a security risk

Fix:
----
If secure-file-priv dir is set, limit gen_dictionary_load() to load files only from
the secure-file-priv dir

### CODE REVIEW

File: plugin/data_masking/src/udf/udf_gen_dictionary_load.cc [hunk 1]
```diff
@@ -95,6 +100,25 @@ static std::string _gen_dictionary_load(const char *dictionary_path,
...
+
+  if (dir_len == 0) {
+    std::stringstream ss;
+    ss << "ERROR: File path is not valid";
+    DBUG_RETURN(ss.str());
+  }
+
+  /* Check if dictionary file is present in --secure_file_priv directory */
+  if (!is_secure_file_path(directory)) {
+    std::stringstream ss;
```
Comment (dutow) [inline plugin/data_masking/src/udf/udf_gen_dictionary_load.cc]: Do we need this stringstream? This error message looks like a fixed string.

File: plugin/data_masking/src/udf/udf_gen_dictionary_load.cc [hunk 2]
```diff
@@ -95,6 +100,25 @@ static std::string _gen_dictionary_load(const char *dictionary_path,
...
   mysql::plugins::tolower(s_dictname);
 
+  /* Extract directory part of the filename */
+  char directory[FN_REFLEN]={0};
+  size_t dir_len = 0;
+  dirname_part(directory, dictionary_path, &dir_len);
+
+  if (dir_len == 0) {
+    std::stringstream ss;
+    ss << "ERROR: File path is not valid";
```
Comment (dutow) [inline plugin/data_masking/src/udf/udf_gen_dictionary_load.cc]: Same question.

File: plugin/data_masking/src/udf/udf_gen_dictionary_load.cc (hunk 1 above)
Comment (satya-bodapati) [reply plugin/data_masking/src/udf/udf_gen_dictionary_load.cc]: Not strictly. Originally I had secure_file_priv printed. The cost is negligible IMO

**Review by dutow — APPROVED** [review]

File: plugin/data_masking/src/udf/udf_gen_dictionary_load.cc (hunk 1 above)
Comment (dutow) [reply plugin/data_masking/src/udf/udf_gen_dictionary_load.cc]: Still, a simple string would be shorter too. Otherwise I agree, this isn't a major issue.

Comment (satya-bodapati) [reply plugin/data_masking/src/udf/udf_gen_dictionary_load.cc]: addressed

File: plugin/data_masking/src/udf/udf_gen_dictionary_load.cc (hunk 2 above)
Comment (satya-bodapati) [reply plugin/data_masking/src/udf/udf_gen_dictionary_load.cc]: addressed

**Review by dutow — APPROVED** [review]

**Comment by satya-bodapati** [comment]
```
~/PS-8.0-7308/bld/mysql-test$ ./mtr --mem --suite=data_masking        
Logging: /home/satya.bodapati/PS-8.0-7308/mysql-test/mysql-test-run.pl  --mem --suite=data_masking
MySQL Version 8.0.21
Checking supported features
 - Binaries are debug compiled
Using suite(s): data_masking
Collecting tests
Checking leftover processes
Removing old var directory
Creating var directory '/home/satya.bodapati/PS-8.0-7308/bld/mysql-test/var'
 - symlinking 'var' to '/dev/shm/var_auto_bMzl'
Installing system database
Using parallel: 1

==============================================================================
                  TEST NAME                       RESULT  TIME (ms) COMMENT
------------------------------------------------------------------------------
[ 10%] data_masking.gen_blacklist                [ pass ]      3
[ 20%] data_masking.gen_dictionary_drop          [ pass ]      3
[ 30%] data_masking.gen_dictionary_load          [ pass ]   9537
[ 40%] data_masking.mask_inner                   [ pass ]      5
[ 50%] data_masking.mask_outer                   [ pass ]      6
[ 60%] data_masking.mask_pan                     [ pass ]      4
[ 70%] data_masking.mask_pan_relaxed             [ pass ]      3
[ 80%] data_masking.mask_ssn                     [ pass ]      4
[ 90%] data_masking.view                         [ pass ]     86
[100%] shutdown_report                           [ pass ]       
------------------------------------------------------------------------------
The servers were restarted 0 times
The servers were reinitialized 0 times
Spent 9.651 of 36 seconds executing testcases

Completed: All 10 tests were successful.
```

**Comment by satya-bodapati** [comment]
https://ps80.cd.percona.com/job/percona-server-8.0-param/779/

**Pull request merged**